Sources/NIOSSL/NIOSSLHandler.swift

```diff
@@ -200,19 +244,56 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
         case .idle, .handshaking, .additionalVerification:
             // we should not flush immediately as we have not completed the handshake and instead buffer the flush
             self.bufferFlush()
-        case .active, .unwrapping, .closing, .unwrapped, .closed:
+        case .active, .unwrapping, .closing, .unwrapped, .inputClosed, .outputClosed, .closed:
             self.bufferFlush()
             self.doUnbufferWrites(context: context)
         }
     }
-    
+
     public func close(context: ChannelHandlerContext, mode: CloseMode, promise: EventLoopPromise<Void>?) {
-        guard mode == .all else {
-            // TODO: Support also other modes ?
+        switch mode {
+        case .output:
+            self.closeOutput(context: context, promise: promise)
+        case .all:
+            self.closeAll(context: context, promise: promise)
+        case .input:
             promise?.fail(ChannelError.operationUnsupported)
-            return
         }
-        
+    }
+
+    private func closeOutput(context: ChannelHandlerContext, promise: EventLoopPromise<Void>?) {
+        switch state {
+        case .closing:
+            // We're in the process of TLS shutdown, which has a higher priority.
+            // Therefore we skip the output closing procedure and cascade the result
+            // of the TLS shutdown request to this new one.
+            if let promise = promise, let closePromise = self.closePromise {
+                closePromise.futureResult.cascade(to: promise)
+            } else if let promise = promise {
+                self.closePromise = promise
+            }
+        case .idle, .outputClosed, .closed, .unwrapping, .unwrapped:
+            // For idle, outputClosed, closed, unwrapping, and unwrapped connections we immediately pass this on to the next
+            // channel handler.
+            context.close(mode: .output, promise: promise)
+        case .active, .handshaking, .additionalVerification, .inputClosed:
+            // It may occur that we are still in the process of handshaking or additional verification. We'll let that happen.
+            // We flush all outstanding writes once the handshake step is complete and set our state to .outputClosed aftwerwards.
+            // This prevents any further writes to this channel.
+            self.state = .outputClosed
+            self.closeOutputPromise = promise
+
+            let flushCompletePromise: EventLoopPromise<Void> = context.eventLoop.makePromise()
+            flushCompletePromise.futureResult.whenComplete { _ in
+                self.doShutdownStep(context: context)
+            }
+            bufferedWritesCascadeFlushPromise(to: flushCompletePromise, context: context)
```

Comment:
So this promise dance is probably unnecessary. We can send `closeOutput` once we've called `flush`, even if the write promises haven't completed. Handlers are required to queue `close(mode: .output)` behind writes.